### PR TITLE
Python: Model parameter with default value as `DefinitionNode`

### PR DESCRIPTION
**Pull request by RasmusWL**
This PR took a series of unexpected events (some could have been avoided with better planning), so the commits end up taking a few twists and turns. I think everything should be very straightforward if reading commit-by-commit :blush:

### CODE REVIEW

**Comment by RasmusWL** [comment]
There might also be an option to fix this problem by using `EssaNodeDefinition` instead, since we have this

https://github.com/github/codeql/blob/d87c8c75d6a988d2564806df94ceb22b61de52e0/python/ql/lib/semmle/python/essa/Essa.qll#L612

But just rewriting the code to be the following didn't work as I wanted. Even though it fixes the missing flow to `x` added in https://github.com/github/codeql/pull/13685/commits/bea07002d380189afd900daa380eab8c27e6c624, it also loses flow to `x_alias` :disappointed: 

```codeql
predicate capturedJumpStep(Node nodeFrom, Node nodeTo) {
  exists(SsaSourceVariable var, EssaNodeDefinition essaDef, ControlFlowNode val |
    essaDef.definedBy(var, val)
  |
    nodeTo.asVar().(ScopeEntryDefinition).getSourceVariable() = var and
    nodeFrom.asCfgNode() = val and
    var.getScope().getScope*() = nodeFrom.getScope()
  )
}
```

I pursued this approach a bit since I wasn't 100% happy about the `.getScope()` checks added to DataFlowPrivate.

I should have added some tests to highlight what the state of dataflow from default values were _before_ this change, since I expect that it's handled to _some_ degree, I just haven't found the logic yet.

**Comment by RasmusWL** [comment]
Clearly I was a bit to eager to just accept changes in 70994b9c573c78169640a8cf9771dfa0eb51bc7c, and now had to fix things up. Keeping that history, since otherwise the changes in 43b0250 might seem strange.

The additions to `ImpliesDataflow.expected` should actually be considered bad, since it selects things that have points-to flow, but no dataflow. However, the tests highlighted below shows that we still have the flow that we want! (so no problem after all)

https://github.com/github/codeql/blob/4b4b9bf9da0d8451e383065008dbf06b4735501b/python/ql/test/experimental/dataflow/coverage/test.py#L856-L860

**Comment by RasmusWL** [comment]
> I pursued this approach a bit since I wasn't 100% happy about the `.getScope()` checks added to DataFlowPrivate.

I think with the current code I'm quite happy. The change below doesn't feel entirely good, but currently I don't see a better approach :neutral_face: 

```diff
   AssignmentDefinition() {
     SsaSource::assignment_definition(this.getSourceVariable(), this.getDefiningNode(), value) and
+    not this instanceof ParameterDefinition
   }
```

> I should have added some tests to highlight what the state of dataflow from default values were _before_ this change, since I expect that it's handled to _some_ degree, I just haven't found the logic yet.

Found the original implementation which I eliminated in 4e8a1144f214901cb6aae55d71d3d4209d1c9122

**Comment by RasmusWL** [comment]
I'm flabbergasted that the previous code did not work.

```diff
   exists(ParameterDefinition param |
     nodeFrom.asCfgNode() = param.getDefault() and
-    nodeTo.asCfgNode() = param.getDefiningNode()
+    nodeTo.asVar() = param.getVariable()
   )
```

How does making the change above make us not find the proper flow??? In the code below, with the fix of this commit we have

`tracked --jumpstep-> CFN param x --local-> ESSA X --> CFN first-use x`

With the change before (and outlined in the diff above) we would have the steps, so CLEARLY that should just work!

* `CFN param x --local-> ESSA X`
* `tracked --jumpstep-> ESSA X`
* `ESSA X --> CFN first-use x`

```py
def foo(x=tracked):
    print(x)
```

it just doesn't make sense to me

**Comment by yoff** [comment]
> I'm flabbergasted that the previous code did not work.
> 
> ```diff
>    exists(ParameterDefinition param |
>      nodeFrom.asCfgNode() = param.getDefault() and
> -    nodeTo.asCfgNode() = param.getDefiningNode()
> +    nodeTo.asVar() = param.getVariable()
>    )
> ```
> 
> How does making the change above make us not find the proper flow??? In the code below, with the fix of this commit we have
> 
> `tracked --jumpstep--> CFN param x --local--> ESSA X --local--> CFN first-use x`
> 
> With the change before (and outlined in the diff above) we would have the steps, so CLEARLY that should just work!
> 
> * `CFN param x --local--> ESSA X`
> * `tracked --jumpstep--> ESSA X`
> * `ESSA X --local--> CFN first-use x`
> 
> ```python
> def foo(x=tracked):
>     print(x)
> ```
> 
> it just doesn't make sense to me

Took me a bit to understand this. The reason is that in your second chain, the link
 
* `tracked --jumpstep--> ESSA X`
 
 is neither followed in the local `flowsTo` part, because it is not a local step (as it is a jump step), nor is it followed in the type tracking part because `ESSA X` is not a local source node. So it is not a valid chain of steps for type tracking. It should be for dataflow, though, but that is not what the test is concerned with.

(In the first case above, `CFN param x` _is_ a local source node and so the chain is valid for type tracking.

File: python/ql/lib/semmle/python/essa/Essa.qll [hunk 1]
```diff
@@ -501,7 +501,8 @@ class AssignmentDefinition extends EssaNodeDefinition {
   ControlFlowNode value;
 
   AssignmentDefinition() {
-    SsaSource::assignment_definition(this.getSourceVariable(), this.getDefiningNode(), value)
+    SsaSource::assignment_definition(this.getSourceVariable(), this.getDefiningNode(), value) and
+    not this instanceof ParameterDefinition
```
Comment (yoff) [inline python/ql/lib/semmle/python/essa/Essa.qll:505]: Consider moving this restriction into `assignment_definition`.

**Review by yoff — COMMENTED** [review]
So, as I understand it, in the following program
```python
def f(x, y=5):
  pass
```
we have two EssaDefinitions, namely for `ESSA x` and `ESSA y`, and their defining nodes (via `getDefiningNode`) are `CFG x` and `CFG y` respectively. This makes sense, since these may receive flow from outside during invocation of `f`.

Additionally, we have a `DefinitionNode`, namely `CFG y`, and its defining node (via `getValue`) is `Expr 5`. This quite nicely represents additional possible flow "from outside" (at definition time, at least) and is mirrored in the newly added jump step.

We do not consider `y = 5` an assignment because it does not happen locally; we would have to split `y` into two or something like that to make it local flow. Did you consider moving this restriction up into `SsaSource::assignment_definition`?

Apart from this one thing, I think it hangs together quite nicely 🙂 

I also checked that
```ql
predicate test(Node node) {
  jumpStepSharedWithTypeTracker(_, node) and
  not node instanceof LocalSourceNode
}
```
is currently empty on two large databases I had lying around.

File: python/ql/lib/semmle/python/essa/Essa.qll (hunk 1 above)
Comment (RasmusWL) [reply python/ql/lib/semmle/python/essa/Essa.qll]: yes, that's better :+1: Done in 98ed5cf52277197024feec05908a8af01aa20ec8

**Review by yoff — APPROVED** [review]
LGTM
We might adjust the comment in the future, though. I think the distinction between parameter definitions and assignment definitions can also be motivated by one being a jump step and one being a local step.

**Pull request merged**